File: fase4/pillar-c/otel-sdk/go/instrumentation.go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/exporters/prometheus"
	"go.opentelemetry.io/otel/sdk/resource"
	"go.opentelemetry.io/otel/sdk/trace"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/semconv/v1.21.0"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// InitializeOTel sets up OpenTelemetry tracing and metrics exporters.
func InitializeOTel(ctx context.Context, serviceName string) (func(context.Context) error, error) {
	// Create resource
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceNameKey.String(serviceName),
			semconv.ServiceVersionKey.String(os.Getenv("VERSION")),
			attribute.String("environment", os.Getenv("ENVIRONMENT")),
			attribute.String("deployment", os.Getenv("DEPLOYMENT")),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create resource: %w", err)
	}

	// Setup tracing with OTLP gRPC exporter
	jaegerHost := os.Getenv("JAEGER_HOST")
	if jaegerHost == "" {
		jaegerHost = "localhost"
	}
	jaegerPort := os.Getenv("JAEGER_PORT")
	if jaegerPort == "" {
		jaegerPort = "4317"
	}

	conn, err := grpc.NewClient(
		fmt.Sprintf("%s:%s", jaegerHost, jaegerPort),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create gRPC connection to Jaeger: %w", err)
	}

	traceExporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, fmt.Errorf("failed to create trace exporter: %w", err)
	}

	tp := trace.NewTracerProvider(
		trace.WithBatcher(traceExporter),
		trace.WithResource(res),
		trace.WithSampler(trace.TraceIDRatioBased(0.1)), // 10% sampling
	)
	otel.SetTracerProvider(tp)

	// Setup metrics with Prometheus exporter
	prometheusExporter, err := prometheus.New()
	if err != nil {
		return nil, fmt.Errorf("failed to create prometheus exporter: %w", err)
	}

	mp := sdkmetric.NewMeterProvider(
		sdkmetric.WithReader(prometheusExporter),
		sdkmetric.WithResource(res),
	)
	otel.SetMeterProvider(mp)

	// Start Prometheus HTTP server
	go func() {
		prometheusPort := os.Getenv("PROMETHEUS_PORT")
		if prometheusPort == "" {
			prometheusPort = "8000"
		}
		http.Handle("/metrics", promhttp.Handler())
		addr := fmt.Sprintf(":%s", prometheusPort)
		log.Printf("Prometheus metrics server started on %s\n", addr)
		if err := http.ListenAndServe(addr, nil); err != nil {
			log.Printf("Prometheus server error: %v\n", err)
		}
	}()

	log.Printf("OpenTelemetry initialized for service: %s\n", serviceName)

	return func(ctx context.Context) error {
		if err := tp.Shutdown(ctx); err != nil {
			return fmt.Errorf("failed to shutdown TracerProvider: %w", err)
		}
		if err := mp.Shutdown(ctx); err != nil {
			return fmt.Errorf("failed to shutdown MeterProvider: %w", err)
		}
		return conn.Close()
	}, nil
}

// GetTracer returns a tracer for the given name.
func GetTracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// GetMeter returns a meter for the given name.
func GetMeter(name string) sdkmetric.Meter {
	return otel.Meter(name)
}

// W3CTraceContext handles W3C TraceContext header propagation.
type W3CTraceContext struct {
	TraceID    string
	SpanID     string
	TraceFlags string
	TraceState string
}

// ExtractTraceContext extracts W3C trace context from HTTP headers.
func ExtractTraceContext(req *http.Request) *W3CTraceContext {
	traceparent := req.Header.Get("traceparent")
	if traceparent == "" {
		return nil
	}

	// Parse traceparent: version-trace_id-parent_id-trace_flags
	var version, traceID, spanID, traceFlags string
	if _, err := fmt.Sscanf(traceparent, "%s-%s-%s-%s", &version, &traceID, &spanID, &traceFlags); err != nil {
		log.Printf("Failed to parse traceparent header: %v\n", err)
		return nil
	}

	if version != "00" {
		return nil // Only support version 00
	}

	return &W3CTraceContext{
		TraceID:    traceID,
		SpanID:     spanID,
		TraceFlags: traceFlags,
		TraceState: req.Header.Get("tracestate"),
	}
}

// InjectTraceContext injects W3C trace context into HTTP headers.
func InjectTraceContext(req *http.Request, ctx *W3CTraceContext) {
	if ctx == nil {
		return
	}
	req.Header.Set("traceparent", fmt.Sprintf("00-%s-%s-%s", ctx.TraceID, ctx.SpanID, ctx.TraceFlags))
	if ctx.TraceState != "" {
		req.Header.Set("tracestate", ctx.TraceState)
	}
}

File: fase4/pillar-c/otel-sdk/python/instrumentation.py
"""
OpenTelemetry instrumentation helper for FastAPI applications.
Provides distributed tracing with OTLP gRPC exporter and Prometheus metrics.
"""

import os
import logging
from typing import Optional

from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.prometheus import PrometheusMetricReader
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor
from opentelemetry.instrumentation.sqlalchemy import SQLAlchemyInstrumentor
from opentelemetry.resources import Resource
from opentelemetry.sdk.resources import SERVICE_NAME
from prometheus_client import start_http_server

# Logging configuration
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


def setup_otel_instrumentation(
    service_name: str,
    jaeger_host: str = "jaeger",
    jaeger_port: int = 4317,
    prometheus_port: int = 8000,
) -> tuple[TracerProvider, MeterProvider]:
    """
    Setup OpenTelemetry instrumentation for distributed tracing and metrics.

    Args:
        service_name: Name of the service (e.g., "manta-gitops")
        jaeger_host: Jaeger collector hostname (default: jaeger)
        jaeger_port: Jaeger OTLP gRPC port (default: 4317)
        prometheus_port: Prometheus metrics scrape port (default: 8000)

    Returns:
        Tuple of (TracerProvider, MeterProvider) for custom instrumentation
    """

    # Create resource
    resource = Resource.create(
        {
            SERVICE_NAME: service_name,
            "environment": os.getenv("ENVIRONMENT", "development"),
            "version": os.getenv("VERSION", "unknown"),
            "deployment": os.getenv("DEPLOYMENT", "kubernetes"),
        }
    )

    # Setup tracing
    otlp_exporter = OTLPSpanExporter(
        endpoint=f"{jaeger_host}:{jaeger_port}",
        insecure=True,  # Set to False in production with proper TLS
        headers=(("Authorization", f"Bearer {os.getenv('OTEL_AUTH_TOKEN', '')}"),)
        if os.getenv("OTEL_AUTH_TOKEN")
        else (),
    )

    trace_provider = TracerProvider(resource=resource)
    trace_provider.add_span_processor(BatchSpanProcessor(otlp_exporter))
    trace.set_tracer_provider(trace_provider)

    # Setup metrics with Prometheus
    prometheus_reader = PrometheusMetricReader()
    metrics_provider = MeterProvider(resource=resource, metric_readers=[prometheus_reader])
    metrics.set_meter_provider(metrics_provider)

    # Start Prometheus HTTP server for metrics scraping
    try:
        start_http_server(port=prometheus_port)
        logger.info(f"Prometheus metrics server started on port {prometheus_port}")
    except OSError as e:
        logger.warning(f"Failed to start Prometheus server: {e}")

    logger.info(f"OpenTelemetry instrumentation initialized for {service_name}")
    return trace_provider, metrics_provider


def instrument_fastapi_app(app) -> None:
    """Instrument a FastAPI application with automatic tracing and metrics."""
    FastAPIInstrumentor.instrument_app(
        app,
        tracer_provider=trace.get_tracer_provider(),
        meter_provider=metrics.get_meter_provider(),
    )
    logger.info("FastAPI application instrumented")


def instrument_requests() -> None:
    """Instrument requests library for HTTP client tracing."""
    RequestsInstrumentor().instrument(tracer_provider=trace.get_tracer_provider())
    logger.info("Requests library instrumented")


def instrument_sqlalchemy(engine) -> None:
    """Instrument SQLAlchemy for database tracing."""
    SQLAlchemyInstrumentor().instrument(
        engine=engine,
        service=os.getenv("SERVICE_NAME", "unknown"),
        tracer_provider=trace.get_tracer_provider(),
    )
    logger.info("SQLAlchemy instrumented")


def get_tracer(name: str) -> trace.Tracer:
    """Get a tracer instance for manual span creation."""
    return trace.get_tracer(__name__)


def get_meter(name: str) -> metrics.Meter:
    """Get a meter instance for manual metric recording."""
    return metrics.get_meter(__name__)


class W3CTraceContextPropagator:
    """
    W3C Trace Context propagator for cross-service span correlation.
    Implements W3C TraceContext specification (https://www.w3.org/TR/trace-context/).
    """

    @staticmethod
    def extract_trace_context(headers: dict) -> Optional[dict]:
        """
        Extract trace context from W3C traceparent header.

        Args:
            headers: HTTP headers dict

        Returns:
            Dict with trace_id, span_id, trace_flags, or None if not present
        """
        traceparent = headers.get("traceparent")
        if not traceparent:
            return None

        try:
            # Format: version-trace_id-parent_id-trace_flags
            parts = traceparent.split("-")
            if len(parts) != 4:
                return None

            version, trace_id, span_id, trace_flags = parts
            if version != "00":  # Only support version 00
                return None

            return {
                "trace_id": trace_id,
                "span_id": span_id,
                "trace_flags": trace_flags,
                "tracestate": headers.get("tracestate", ""),
            }
        except Exception as e:
            logger.warning(f"Failed to extract trace context: {e}")
            return None

    @staticmethod
    def inject_trace_context(span_context: dict) -> dict:
        """
        Inject trace context into W3C traceparent header.

        Args:
            span_context: Dict with trace_id, span_id, trace_flags

        Returns:
            Dict with traceparent and tracestate headers
        """
        trace_id = span_context.get("trace_id", "0" * 32)
        span_id = span_context.get("span_id", "0" * 16)
        trace_flags = span_context.get("trace_flags", "01")
        tracestate = span_context.get("tracestate", "")

        return {
            "traceparent": f"00-{trace_id}-{span_id}-{trace_flags}",
            "tracestate": tracestate,
        }
